linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2213,15 +2247,25 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
         assert(isStat, "Found Assign in expression position")
         assert(optReceiver.isDefined,
             "There was a This(), there should be a receiver")
-        pretransformSelectCommon(lhs.tpe, optReceiver.get._2, className, field,
-            isLhsOfAssign = true) { preTransLhs =>
-          // TODO Support assignment of record
-          cont(PreTransTree(
-              Assign(finishTransformExpr(preTransLhs).asInstanceOf[AssignLhs],
-                  finishTransformExpr(args.head)),
-              RefinedType.NoRefinedType))
+
+        val trhs = args.head
+
+        val tree = {
+          if (!isFieldRead(className, field.name)) {
+            // Field is never read, discard assign, keep side effects only.
+            finishTransformStat(trhs)
+          } else {
+            trampoline {
+              pretransformSelectCommon(lhs.tpe, optReceiver.get._2, className, field,
+                  isLhsOfAssign = true) { tlhs =>
+                finishTransformAssign(tlhs, args.head)
+              }
+            }
```

Comment:
I don't think the `trampoline` here is necessary. You can push the `cont(...)` from below in the two branches and get rid of it. Something like:
```suggestion
          if (!isFieldRead(className, field.name)) {
            // Field is never read, discard assign, keep side effects only.
            cont(PreTransTree(finishTransformStat(trhs)))
          } else {
            // trampoline {
              pretransformSelectCommon(lhs.tpe, optReceiver.get._2, className, field,
                  isLhsOfAssign = true) { tlhs =>
                cont(PreTransTree(finishTransformAssign(tlhs, args.head)))
              }
            // }
```
There's not even a need for the `NoRefinedType`, and it might be counter-productive if `finishTransform{Stat,Assign}` collapses to a tree of type `NothingType`.